src/app/reporting/Engine.cpp

```diff
@@ -336,6 +367,65 @@ void Engine::Run()
         mCurReadHandlerIdx = (mCurReadHandlerIdx + 1) % CHIP_IM_MAX_NUM_READ_HANDLER;
         readHandler        = imEngine->mReadHandlers + mCurReadHandlerIdx;
     }
+    for (auto & path : mChangeSetPathList)
+    {
+        path.ClearDirty();
+    }
+}
+
+void Engine::SetDirty(ClusterInfo & aClusterInfo)
+{
+    for (auto & path : mChangeSetPathList)
+    {
+        if (path.IsDirty())
+        {
+            // If aClusterInfo is part of existing dirty path, skip it
+            // If aClusterInfo is super set of one existing path, update that path with new path.
+            if (path.IsAttributePathSupersetOf(aClusterInfo))
+            {
+                return;
+            }
+            if (aClusterInfo.IsAttributePathSupersetOf(path))
+            {
+                path.mListIndex = aClusterInfo.mListIndex;
+                path.mFieldId   = aClusterInfo.mFieldId;
+                path.mFlags     = aClusterInfo.mFlags;
+                return;
+            }
+        }
+    }
+
+    for (auto & readHandler : InteractionModelEngine::GetInstance()->mReadHandlers)
+    {
+        if (readHandler.IsSubscription() && (readHandler.IsReportable() || readHandler.IsReporting()))
+        {
+            ClusterInfo * clusterInstance = readHandler.GetAttributeClusterInfolist();
+            while (clusterInstance != nullptr)
+            {
+                // If dirty path intersects with interested path from handler, it mark this in ChangeSetPathList and also
+                // mark cluster as dirty for the particular subscribe handler , otherwise, it skip this dirty path.
+                if (clusterInstance->IsAttributePathSupersetOf(aClusterInfo) ||
+                    aClusterInfo.IsAttributePathSupersetOf(*clusterInstance))
+                {
+                    for (auto & path : mChangeSetPathList)
+                    {
+                        if (!path.IsDirty())
+                        {
+                            path = aClusterInfo;
+                            path.SetDirty();
+                            clusterInstance->SetDirty();
+                            ChipLogProgress(DataManagement, "<RE> SetDiry Success");
+                            return;
+                        }
+                    }
+                    ChipLogProgress(DataManagement, "<RE> SetDirty fails with no available clean path in mChangeSetPathList");
```

Comment:
We need a graceful fall back story, since just returning and not marking the path as dirty isn't going to be cut it, since no reports will get sent indefinitely for that dirty item till it changes again next time.

Not to be addressed here, but needs a follow-up issue.

Reply:
updated.